### ADS 509
### Assignment 1.1
### Dingyi Duan
#### GitHub: https://github.com/dingyiduan7/ADS_509_module1

# ADS 509 Module 1: APIs and Web Scraping

This notebook has three parts. In the first part you will pull data from the Twitter API. In the second, you will scrape lyrics from AZLyrics.com. In the last part, you'll run code that verifies the completeness of your data pull. 

For this assignment you have chosen two musical artists who have at least 100,000 Twitter followers and 20 songs with lyrics on AZLyrics.com. In this part of the assignment we pull the some of the user information for the followers of your artist and store them in text files. 


## Important Note

This assignment requires you to have a version of Tweepy that is at least version 4. The latest version is 4.10 as I write this. Critically, this version of Tweepy is *not* on the upgrade path from Version 3, so you will not be able to simply upgrade the package if you are on Version 3. Instead you will need to explicitly install version 4, which you can do with a command like this: `pip install "tweepy>=4"`. You will also be using Version 2 of the Twitter API for this assignment. 

Run the below cell. If your version of Tweepy begins with a "4", then you should be good to go. If it begins with a "3" then run the following command, found [here](https://stackoverflow.com/questions/5226311/installing-specific-package-version-with-pip), at the command line or in a cell: `pip install -Iv tweepy==4.9`. (You may want to update that version number if Tweepy has moved on past 4.9. 

In [1]:
#!pip show tweepy
#!pip install -Iv tweepy==4.9

## General Assignment Instructions

These instructions are included in every assignment, to remind you of the coding standards for the class. Feel free to delete this cell after reading it. 

One sign of mature code is conforming to a style guide. We recommend the [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html). If you use a different style guide, please include a cell with a link. 

Your code should be relatively easy-to-read, sensibly commented, and clean. Writing code is a messy process, so please be sure to edit your final submission. Remove any cells that are not needed or parts of cells that contain unnecessary code. Remove inessential `import` statements and make sure that all such statements are moved into the designated cell. 

Make use of non-code cells for written commentary. These cells should be grammatical and clearly written. In some of these cells you will have questions to answer. The questions will be marked by a "Q:" and will have a corresponding "A:" spot for you. *Make sure to answer every question marked with a `Q:` for full credit.* 


# Twitter API Pull

In [2]:
# for the twitter section
import tweepy
import os
import datetime
import re
from pprint import pprint

# for the lyrics scrape section
import requests
import time
from bs4 import BeautifulSoup
from collections import defaultdict, Counter


In [45]:
# Use this cell for any import statements you add

import random
import os.path
import re
import pandas as pd

We need bring in our API keys. Since API keys should be kept secret, we'll keep them in a file called `api_keys.py`. This file should be stored in the directory where you store this notebook. The example file is provided for you on Blackboard. The example has API keys that are _not_ functional, so you'll need to get Twitter credentials and replace the placeholder keys. 

In [4]:
from api_keys import api_key, api_key_secret, bearer_token

In [5]:
client = tweepy.Client(bearer_token,wait_on_rate_limit=True)

# Testing the API

The Twitter APIs are quite rich. Let's play around with some of the features before we dive into this section of the assignment. For our testing, it's convenient to have a small data set to play with. We will seed the code with the handle of John Chandler, one of the instructors in this course. His handle is `@37chandler`. Feel free to use a different handle if you would like to look at someone else's data. 

We will write code to explore a few aspects of the API: 

1. Pull some of the followers @37chandler.
1. Explore response data, which gives us information about Twitter users. 
1. Pull the last few tweets by @37chandler.


In [6]:
handle = "37chandler"
user_obj = client.get_user(username=handle)

followers = client.get_users_followers(
    # Learn about user fields here: 
    # https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/user
    user_obj.data.id, user_fields=["created_at","description","location",
                                   "public_metrics"]
)



Now let's explore these a bit. We'll start by printing out names, locations, following count, and followers count for these users. 

In [7]:
num_to_print = 20

for idx, user in enumerate(followers.data) :
    following_count = user.public_metrics['following_count']
    followers_count = user.public_metrics['followers_count']
    
    print(f"{user.name} lists '{user.location}' as their location.")
    print(f" Following: {following_count}, Followers: {followers_count}.")
    print()
    
    if idx >= (num_to_print - 1) :
        break
    

Dave Renn lists 'None' as their location.
 Following: 42, Followers: 10.

Lionel lists 'None' as their location.
 Following: 202, Followers: 204.

Megan Randall lists 'None' as their location.
 Following: 141, Followers: 100.

Jacob Salzman lists 'None' as their location.
 Following: 562, Followers: 134.

twiter not fun lists 'None' as their location.
 Following: 221, Followers: 21.

Hariettwilsonincarnate lists 'None' as their location.
 Following: 218, Followers: 60.

Christian Tinsley lists 'None' as their location.
 Following: 2, Followers: 0.

Steve lists 'I'm over here.' as their location.
 Following: 1590, Followers: 33.

John O'Connor 🇺🇦 lists 'None' as their location.
 Following: 8, Followers: 1.

CodeGrade lists 'Amsterdam' as their location.
 Following: 2820, Followers: 422.

Cleverhood lists 'Providence, RI' as their location.
 Following: 2792, Followers: 3558.

Regina 🚶‍♀️🚲🌳 lists 'Minneapolis' as their location.
 Following: 2796, Followers: 3321.

Eric Hallstrom lists 'Mi

Let's find the person who follows this handle who has the most followers. 

In [8]:
max_followers = 0

for idx, user in enumerate(followers.data) :
    followers_count = user.public_metrics['followers_count']
    
    if followers_count > max_followers :
        max_followers = followers_count
        max_follower_user = user

        
print(max_follower_user)
print(max_follower_user.public_metrics)

WedgeLIVE
{'followers_count': 14167, 'following_count': 2223, 'tweet_count': 56086, 'listed_count': 218}


Let's pull some more user fields and take a look at them. The fields can be specified in the `user_fields` argument. 

In [9]:
response = client.get_user(id=user_obj.data.id,
                          user_fields=["created_at","description","location",
                                       "entities","name","pinned_tweet_id","profile_image_url",
                                       "verified","public_metrics"])

In [10]:
for field, value in response.data.items() :
    print(f"for {field} we have {value}")

for description we have He/Him. Data scientist, urban cyclist, educator, erstwhile frisbee player. 

¯\_(ツ)_/¯
for username we have 37chandler
for verified we have False
for location we have MN
for name we have John Chandler
for id we have 33029025
for created_at we have 2009-04-18 22:08:22+00:00
for profile_image_url we have https://pbs.twimg.com/profile_images/2680483898/b30ae76f909352dbae5e371fb1c27454_normal.png
for public_metrics we have {'followers_count': 193, 'following_count': 589, 'tweet_count': 996, 'listed_count': 3}


Now a few questions for you about the user object.

Q: How many fields are being returned in the `response` object? 

A: 9 fields

---

Q: Are any of the fields within the user object non-scalar? (I.e., more complicated than a simple data type like integer, float, string, boolean, etc.) 

A: Yes, for public_metrics - it's a dictionary rather than a simple data type. 

---

Q: How many friends, followers, and tweets does this user have? 

A: friends: 589; followers: 193; tweets: 994. 


Although you won't need it for this assignment, individual tweets can be a rich source of text-based data. To illustrate the concepts, let's look at the last few tweets for this user. You are encouraged to explore the fields that are available about Tweets.

In [11]:

response = client.get_users_tweets(user_obj.data.id)

# By default, only the ID and text fields of each Tweet will be returned
for idx, tweet in enumerate(response.data) :
    print(tweet.id)
    print(tweet.text)
    print()
    
    if idx > 10 :
        break

1569155273742327811
As a Minneapolis person, I knew we had Toronto beat, but I didn't realize Portland had us beat: https://t.co/xrx5mOFcWK.

But @nytimes, c'mon! https://t.co/M9mBWhdgsj

1568982292923826176
RT @wonderofscience: Amazing lenticular cloud over Mount Fuji

Credit: Iurie Belegurschi
https://t.co/0mUxl28H9U

1568242374085869570
RT @depthsofwiki: lots of memes about speedy wikipedia editors — quick thread about what went down on wikipedia in the minutes after her de…

1568074978754703361
@DrLaurenWilson @leighradwood @MaritsaGeorgiou @Walgreens I could not possibly agree more with this sentiment. Compared to almost any other primary care I've received, they are great.

1567530169686196224
@DrLaurenWilson @MaritsaGeorgiou @Walgreens For those who have access to Curry Health Center on campus, you can get a bivalent booster in 15 minutes from their delightful staff.

1567511181526708224
RT @shes_the_maNN1: I can’t describe how ancient this makes me feel. https://t.co/a1IvELjOFY

## Pulling Follower Information

In this next section of the assignment, we will pull information about the followers of your two artists. We've seen above how to pull a set of followers using `client.get_users_followers`. This function has a parameter, `max_results`, that we can use to change the number of followers that we pull. Unfortunately, we can only pull 1000 followers at a time, which means we will need to handle the _pagination_ of our results. 

The return object has the `.data` field, where the results will be found. It also has `.meta`, which we use to select the next "page" in the results using the `next_token` result. I will illustrate the ideas using our user from above. 


### Rate Limiting

Twitter limits the rates at which we can pull data, as detailed in [this guide](https://developer.twitter.com/en/docs/twitter-api/rate-limits). We can make 15 user requests per 15 minutes, meaning that we can pull $4 \cdot 15 \cdot 1000 = 60000$ users per hour. I illustrate the handling of rate limiting below, though whether or not you hit that part of the code depends on your value of `handle`.  


In the below example, I'll pull all the followers, 25 at a time. (We're using 25 to illustrate the idea; when you do this set the value to 1000.) 

In [12]:
handle_followers = []
pulls = 0
max_pulls = 10
next_token = None

while True :

    followers = client.get_users_followers(
        user_obj.data.id, 
        max_results=25, # when you do this for real, set this to 1000!
        pagination_token = next_token,
        user_fields=["created_at","description","location",
                     "entities","name","pinned_tweet_id","profile_image_url",
                     "verified","public_metrics"]
    )
    pulls += 1
    
    for follower in followers.data : 
        follower_row = (follower.id,follower.name,follower.created_at,follower.description)
        handle_followers.append(follower_row)
    
    if 'next_token' in followers.meta and pulls < max_pulls :
        next_token = followers.meta['next_token']
    else : 
        break
        



## Pulling Twitter Data for Your Artists

Now let's take a look at your artists and see how long it is going to take to pull all their followers. 

In [13]:
artists = dict()

for handle in ['robynkonichiwa','cher'] : 
    user_obj = client.get_user(username=handle,user_fields=["public_metrics"])
    artists[handle] = (user_obj.data.id, 
                       handle,
                       user_obj.data.public_metrics['followers_count'])
    

for artist, data in artists.items() : 
    print(f"It would take {data[2]/(1000*15*4):.2f} hours to pull all {data[2]} followers for {artist}. ")
    

It would take 5.94 hours to pull all 356363 followers for robynkonichiwa. 
It would take 67.42 hours to pull all 4045320 followers for cher. 


Depending on what you see in the display above, you may want to limit how many followers you pull. It'd be great to get at least 200,000 per artist. 

As we pull data for each artist we will write their data to a folder called "twitter", so we will make that folder if needed.

In [14]:
# Make the "twitter" folder here. If you'd like to practice your programming, add functionality 
# that checks to see if the folder exists. If it does, then "unlink" it. Then create a new one.

if not os.path.isdir("twitter") : 
    #shutil.rmtree("twitter/")
    os.mkdir("twitter")

In this following cells, build on the above code to pull some of the followers and their data for your two artists. As you pull the data, write the follower ids to a file called `[artist name]_followers.txt` in the "twitter" folder. For instance, for Cher I would create a file named `cher_followers.txt`. As you pull the data, also store it in an object like a list or a data frame.

In addition to creating a file that only has follower IDs in it, you will create a file that includes user data. From the response object please extract and store the following fields: 

* screen_name	
* name	
* id	
* location	
* followers_count	
* friends_count	
* description

Store the fields with one user per row in a tab-delimited text file with the name `[artist name]_follower_data.txt`. For instance, for Cher I would create a file named `cher_follower_data.txt`. 

One note: the user's description can have tabs or returns in it, so make sure to clean those out of the description before writing them to the file. I've included some example code to do that below the stub. 

In [ ]:
save_path = 'D:/USD-ADS_graduate/ADS-509/Module 1/ADS_509_module1/twitter'
handles = ['MariahCarey','NICKIMINAJ']
max_pulls = 101
whitespace_pattern = re.compile(r"\s+")

"""
I have changed the variable names to  'followers_id' and 'followers_data' for more intuitive reading
All relevant names are also changed 
"""
followers_id = dict()
followers_data = dict() 

# Grabs the time when we start making requests to the API
start_time = datetime.datetime.now()

for handle in handles :
    
    followers_id[handle] = [] # will be a simple list of IDs
    followers_data[handle] = [] # will be a list of lists
    user_obj = client.get_user(username=handle)
    pulls = 0
    follower_count = 0
    header_list = ['screen_name', 'name', 'location','description','followers_count','friends_count']
    
    # Create the output file names 
    
    followers_id_file = handle + "_followers_id.txt"
    followers_data_file = handle + "_followers.txt"
    
    for followers in tweepy.Paginator(client.get_users_followers,
                                      user_obj.data.id,
                                      max_results = 999,
                                      limit = 101,
                                      user_fields=["username","name","id","location",
                                                   "description","public_metrics"]
                                     ):
    
        for follower in followers.data : 
            
            # append followers'id to the list
            followers_id[handle].append(follower.id) 
            
            # set followers_id file path
            followers_id_path = os.path.join(save_path, followers_id_file)
            
            
            
            
            # append followers' data to the list
            followers_data_row = (str(follower.username),
                                  str(follower.name),
                                  str(follower.location),
                                  re.sub(r"\s+"," ",follower.description).strip(), # clean description
                                  str(follower.public_metrics['followers_count']),
                                  str(follower.public_metrics['following_count'])
                                 )
            followers_data[handle].append(followers_data_row)
            
            
            # set followers_data file path
            followers_data_path = os.path.join(save_path, followers_data_file)
            
        print(pulls+1, "pulls completed for", handle)
        pulls += 1
        follower_count += len(followers_id[handle])
        print("Total of", len(followers_id[handle]), "followers pulled for", handle)
            
        if 'next_token' in followers.meta and pulls < max_pulls :
            next_token = followers.meta['next_token']
        else : 
            break
    
    # save data to dataframes and write to csvs, seprated by tab
    pd.DataFrame(followers_id[handle]).to_csv(followers_id_path, index=False, sep='\t',header = False)
    print(follower_count,'followers IDs have been written to %s'%followers_id_file)
    pd.DataFrame(followers_data[handle]).to_csv(followers_data_path, index=False, sep='\t', header=header_list)
    print(follower_count,'followers data has been written to %s'%followers_data_file)

    
# Let's see how long it took to grab all follower IDs
end_time = datetime.datetime.now()
print("Total time consumed is:", end_time-start_time)


1 pulls completed for MariahCarey
Total of 999 followers pulled for MariahCarey
2 pulls completed for MariahCarey
Total of 1998 followers pulled for MariahCarey
3 pulls completed for MariahCarey
Total of 2997 followers pulled for MariahCarey
4 pulls completed for MariahCarey
Total of 3996 followers pulled for MariahCarey
5 pulls completed for MariahCarey
Total of 4995 followers pulled for MariahCarey
6 pulls completed for MariahCarey
Total of 5993 followers pulled for MariahCarey
7 pulls completed for MariahCarey
Total of 6992 followers pulled for MariahCarey
8 pulls completed for MariahCarey
Total of 7990 followers pulled for MariahCarey
9 pulls completed for MariahCarey
Total of 8989 followers pulled for MariahCarey
10 pulls completed for MariahCarey
Total of 9988 followers pulled for MariahCarey
11 pulls completed for MariahCarey
Total of 10987 followers pulled for MariahCarey
12 pulls completed for MariahCarey
Total of 11986 followers pulled for MariahCarey
13 pulls completed for M

Rate limit exceeded. Sleeping for 891 seconds.


15 pulls completed for MariahCarey
Total of 14983 followers pulled for MariahCarey


In [26]:
tricky_description = """
    Home by Warsan Shire
    
    no one leaves home unless
    home is the mouth of a shark.
    you only run for the border
    when you see the whole city
    running as well.

"""
# This won't work in a tab-delimited text file.

clean_description = re.sub(r"\s+"," ",tricky_description).strip()
clean_description

'Home by Warsan Shire no one leaves home unless home is the mouth of a shark. you only run for the border when you see the whole city running as well.'

---

# Lyrics Scrape

This section asks you to pull data from the Twitter API and scrape www.AZLyrics.com. In the notebooks where you do that work you are asked to store the data in specific ways. 

In [88]:
artists = {'Mariah Carey':"https://www.azlyrics.com/m/mariahcarey.html",
           'Nicki Minaj':"https://www.azlyrics.com/n/nickiminaj.html"} 
# we'll use this dictionary to hold both the artist name and the link on AZlyrics

## A Note on Rate Limiting

The lyrics site, www.azlyrics.com, does not have an explicit maximum on number of requests in any one time, but in our testing it appears that too many requests in too short a time will cause the site to stop returning lyrics pages. (Entertainingly, the page that gets returned seems to only have the song title to [a Tom Jones song](https://www.azlyrics.com/lyrics/tomjones/itsnotunusual.html).) 

Whenever you call `requests.get` to retrieve a page, put a `time.sleep(5 + 10*random.random())` on the next line. This will help you not to get blocked. If you _do_ get blocked, which you can identify if the returned pages are not correct, just request a lyrics page through your browser. You'll be asked to perform a CAPTCHA and then your requests should start working again. 

## Part 1: Finding Links to Songs Lyrics

That general artist page has a list of all songs for that artist with links to the individual song pages. 

Q: Take a look at the `robots.txt` page on www.azlyrics.com. (You can read more about these pages [here](https://developers.google.com/search/docs/advanced/robots/intro).) Is the scraping we are about to do allowed or disallowed by this page? How do you know? 

A: By looking at https://www.azlyrics.com/robots.txt, we can see that scraping under the lyrics database (/lyricsdb/) and the songs (/song/) are NOT allowed.


In [ ]:
# Let's set up a dictionary of lists to hold our links
lyrics_pages = dict()

for artist, artist_page in artists.items() :
    # request the page and sleep
    r = requests.get(artist_page)
    soup = BeautifulSoup(r.text, 'lxml')
    time.sleep(5 + 10*random.random())
    url = 'https://www.azlyrics.com'
    lyrics_pages[artist] = [url+a['href'].strip('..') for a in soup.find(id='listAlbum').findAll('a', href=True)]
    
    # the line above is from: https://stackoverflow.com/questions/61101772/scraping-a-webpage-and-related-subsequent-page-using-r-or-python
    

    # now extract the links to lyrics pages from this page
    # store the links `lyrics_pages` where the key is the artist and the
    # value is a list of links. 
    

Let's make sure we have enough lyrics pages to scrape. 

In [ ]:
for artist, lp in lyrics_pages.items() :
    assert(len(set(lp)) > 20) 

In [ ]:
# Let's see how long it's going to take to pull these lyrics 
# if we're waiting `5 + 10*random.random()` seconds 
for artist, links in lyrics_pages.items() : 
    print(f"For {artist} we have {len(links)}.")
    print(f"The full pull will take for this artist will take {round(len(links)*10/3600,2)} hours.")

## Part 2: Pulling Lyrics

Now that we have the links to our lyrics pages, let's go scrape them! Here are the steps for this part. 

1. Create an empty folder in our repo called "lyrics". 
1. Iterate over the artists in `lyrics_pages`. 
1. Create a subfolder in lyrics with the artist's name. For instance, if the artist was Cher you'd have `lyrics/cher/` in your repo.
1. Iterate over the pages. 
1. Request the page and extract the lyrics from the returned HTML file using BeautifulSoup.
1. Use the function below, `generate_filename_from_url`, to create a filename based on the lyrics page, then write the lyrics to a text file with that name. 


In [ ]:
def generate_filename_from_link(link) :
    
    if not link :
        return None
    
    # drop the http or https and the html
    name = link.replace("https","").replace("http","")
    name = link.replace(".html","")

    name = name.replace("/lyrics/","")
    
    # Replace useless chareacters with UNDERSCORE
    name = name.replace("://","").replace(".","_").replace("/","_")
    
    # tack on .txt
    name = name + ".txt"
    
    return(name)


In [ ]:
# Make the lyrics folder here. If you'd like to practice your programming, add functionality 
# that checks to see if the folder exists. If it does, then use shutil.rmtree to remove it and create a new one.

if not os.path.isdir("lyrics") : 
    #shutil.rmtree("lyrics/")

    os.mkdir("lyrics")

In [ ]:
url_stub = "https://www.azlyrics.com" 
start = time.time()
total_pages = 25  # set max file condition

for artist in lyrics_pages :
    
    start_time2 = datetime.datetime.now()
    
    # Use this space to carry out the following steps: 
    
    # 1. Build a subfolder for the artist
    # 2. Iterate over the lyrics pages
    # 3. Request the lyrics page. 
        # Don't forget to add a line like `time.sleep(5 + 10*random.random())`
        # to sleep after making the request
    # 4. Extract the title and lyrics from the page.
    # 5. Write out the title, two returns ('\n'), and the lyrics. Use `generate_filename_from_url`
    #    to generate the filename. 
    
    # Remember to pull at least 20 songs per artist. It may be fun to pull all the songs for the artist
    
    pages = 0 # start with 0 file
    
    """
    The code below create a folder under /lyrics for each artist
    """
    
    directory = artist
    # Parent Directory path
    parent_dir = "D:/USD-ADS_graduate/ADS-509/Module 1/ADS_509_module1/lyrics"
    # Path
    artist_dir = os.path.join(parent_dir, directory)
    if not os.path.isdir(artist_dir) :
        os.mkdir(artist_dir)
    
    """
    The for loop below is to iterate through all the lyrics links per artist
    """
    for link in lyrics_pages[artist] :
        
        name = generate_filename_from_link(link) # get the name of the file for each lyric
        
        r2 = requests.get(link) # send request
        soup2 = BeautifulSoup(r2.content, 'lxml') # use beautiful soup to retreive content  
        lyric = soup2.select_one(".ringtone ~ div").get_text(strip=True, separator="\n")
        
        """
        the two lines below uses regx to extract the title from the link
        """
        s = re.search('lyrics/(.*).html', link).group(1)
        title = re.findall(r'/(\w+)', s)
        
        # write the title and lyric to a file per artist
        lyric_file_path = os.path.join(artist_dir, name)
        with open(lyric_file_path, 'w') as fp3:
            # write title and lyrics
            fp3.write("%s\n\n%s" % (title,lyric))
        
        # add one iteration and sleep for some time
        pages += 1
        time.sleep(5 + 10*random.random())
        
        # breaks the loop when enough files are created
        if pages > total_pages:
            break
            
end_time2 = datetime.datetime.now()
print("Total time consumed is:", end_time2-start_time2) # print the time consumed

---

# Evaluation

This assignment asks you to pull data from the Twitter API and scrape www.AZLyrics.com.  After you have finished the above sections , run all the cells in this notebook. Print this to PDF and submit it, per the instructions.

In [ ]:
# Simple word extractor from Peter Norvig: https://norvig.com/spell-correct.html
def words(text): 
    return re.findall(r'\w+', text.lower())

---

## Checking Twitter Data

The output from your Twitter API pull should be two files per artist, stored in files with formats like `cher_followers.txt` (a list of all follower IDs you pulled) and `cher_followers_data.txt`. These files should be in a folder named `twitter` within the repository directory. This code summarizes the information at a high level to help the instructor evaluate your work. 

In [ ]:
twitter_files = os.listdir("twitter")
twitter_files = [f for f in twitter_files if f != ".DS_Store"]
artist_handles = list(set([name.split("_")[0] for name in twitter_files]))

print(f"We see two artist handles: {artist_handles[0]} and {artist_handles[1]}.")

In [ ]:
for artist in artist_handles :
    followers_id_file = artist + "_followers_id.txt"
    followers_data_file = artist + "_followers.txt"

    
    ids = open("twitter/" + followers_id_file,'r', encoding="utf-8").readlines()
    
    print(f"We see {len(ids)-1} in your follower file for {artist}, assuming a header row.")
    
    with open("twitter/" + followers_data_file,'r', encoding="utf-8") as infile :
        
        # check the headers
        headers = infile.readline().split("\t")
        
        print(f"In the follower data file ({followers_data_file}) for {artist}, we have these columns:")
        print(" : ".join(headers))
        
        description_words = []
        locations = set()
        
        
        for idx, line in enumerate(infile.readlines()) :
            line = line.strip("\n").split("\t")
            
            try : 
                locations.add(line[2]) # change the index number to match the right field           
                description_words.extend((line[3])) # change the index number to match the right field   
            except :
                pass
    
        

        print(f"We have {idx+1} data rows for {artist} in the follower data file.")

        print(f"For {artist} we have {len(locations)} unique locations.")

        print(f"For {artist} we have {len(description_words)} words in the descriptions.")
        print("Here are the five most common words:")
        print(Counter(description_words).most_common(5))

        
        print("")
        print("-"*40)
        print("")
    

## Checking Lyrics 

The output from your lyrics scrape should be stored in files located in this path from the directory:
`/lyrics/[Artist Name]/[filename from URL]`. This code summarizes the information at a high level to help the instructor evaluate your work. 

In [ ]:
artist_folders = os.listdir("lyrics/")
artist_folders = [f for f in artist_folders if os.path.isdir("lyrics/" + f)]

for artist in artist_folders : 
    artist_files = os.listdir("lyrics/" + artist)
    artist_files = [f for f in artist_files if 'txt' in f or 'csv' in f or 'tsv' in f]

    print(f"For {artist} we have {len(artist_files)} files.")

    artist_words = []

    for f_name in artist_files : 
        with open("lyrics/" + artist + "/" + f_name) as infile : 
            artist_words.extend(words(infile.read()))

            
    print(f"For {artist} we have roughly {len(artist_words)} words, {len(set(artist_words))} are unique.")
